### PR TITLE
PLUGIN-1089: Improve ServiceNow Plugins Performance

### PR DESCRIPTION
**Pull request by sgarg-CS**
[PLUGIN-1089](https://cdap.atlassian.net/browse/PLUGIN-1089) 

Add retry to ServiceNow RestAPI Calls when you get the maximum execution time exceeded error or `OauthSystemException`

### CODE REVIEW

**Comment by sgarg-CS** [comment]
hi @rmstar 

Can you please review this PR?

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 1]
```diff
@@ -48,6 +50,7 @@
   private int httpStatus;
   private Map<String, String> headers;
   private String responseBody;
+  private boolean isRetriable;
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: rename to retryable

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 2]
```diff
@@ -98,7 +101,24 @@ public int getHttpStatus() {
...
+    } else if (successCodes.contains(getHttpStatus())) {
+      isSuccess = true;
+    }
+    return isSuccess;
+  }
+
+  private boolean checkRetriable() {
+    Gson gson = new Gson();
+    JsonObject jo = gson.fromJson(this.getResponseBody(), JsonObject.class);
+    if (jo.get("status") != null && jo.get("status").getAsString().equals("failure")) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: define keywords like status failure or error message as String literals and use them.
Merge these two for loops into one.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java [hunk 3]
```diff
@@ -78,23 +78,25 @@ public StructuredRecord getCurrentValue() throws IOException {
...
     // Get the table data
-    results =
-      restApi.fetchTableRecords(tableName, multiSourcePluginConf.getStartDate(), multiSourcePluginConf.getEndDate(),
-                                split.getOffset(), ServiceNowConstants.PAGE_SIZE);
-
-    LOG.debug("size={}", results.size());
+    results = restApi.fetchTableRecordsRetriableMode(tableName, multiSourcePluginConf.getStartDate(),
+      multiSourcePluginConf.getEndDate(), split.getOffset(), ServiceNowConstants.PAGE_SIZE, false);
     if (!results.isEmpty()) {
+      LOG.debug("size={}", results.size());
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java]: write a more clear message for user to understand.

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 4]
```diff
@@ -48,7 +49,14 @@
   private static final String FIELD_CREATED_ON = "sys_created_on";
   private static final String FIELD_UPDATED_ON = "sys_updated_on";
   private static final String OAUTH_URL_TEMPLATE = "%s/oauth_token.do";
-
+  private static final Map<Integer, Integer> retryDelay = new HashMap<Integer, Integer>() {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: make it immutable.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 1 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Renamed to `retryable` here and at other places. a752f6b05f68ae64ce9ed63c94fc7b1cc1865b12

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 2 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Defined String Literals and merged two **if** blocks a752f6b05f68ae64ce9ed63c94fc7b1cc1865b12

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java (hunk 3 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java]: Re-written the debug logs that print the size of the result. a752f6b05f68ae64ce9ed63c94fc7b1cc1865b12

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 4 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Replaced the HashMap with ImmutableMap. a752f6b05f68ae64ce9ed63c94fc7b1cc1865b12

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 5]
```diff
@@ -48,7 +49,11 @@
   private static final String FIELD_CREATED_ON = "sys_created_on";
   private static final String FIELD_UPDATED_ON = "sys_updated_on";
   private static final String OAUTH_URL_TEMPLATE = "%s/oauth_token.do";
-
+  private static final ImmutableMap<Integer, Integer> retryDelay = ImmutableMap.of(
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Use a library instead of implementing retry/backoff logic yourself.
Example - https://github.com/data-integrations/google-cloud/pull/918/files#diff-6a7f8b65e9e06b9f3d93abd2603ec319c4e9c5b6083fcadac773b32f55e1da8cR318

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 6]
```diff
@@ -89,14 +94,23 @@ public String getAccessToken() throws OAuthSystemException, OAuthProblemExceptio
     try {
       String accessToken = getAccessToken();
       requestBuilder.setAuthHeader(accessToken);
+      LOG.debug("Request URL: {} ", requestBuilder.build().getUrl());
+      long start = System.currentTimeMillis();
       apiResponse = executeGet(requestBuilder.build());
+      long end = System.currentTimeMillis();
+      LOG.debug("restAPI execution time for {} to {} records took {}s ", offset,
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: we can remove logging added for debugging purposes

Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: we can remove logging added for debugging purposes

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 7]
```diff
@@ -98,7 +102,25 @@ public int getHttpStatus() {
   }
 
   public boolean isSuccess() {
-    return successCodes.contains(getHttpStatus());
+    boolean isSuccess = false;
+    if (checkRetryable()) {
+      isSuccess = false;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: why do we need need this if (checkRetryable()) check?  Can't we just return true if successCodes.contains(getHttpStatus()) - i.e. what this method was doing previously

Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: why do we need need this if (checkRetryable()) check?  Can't we just return true if successCodes.contains(getHttpStatus()) - i.e. what this method was doing previously

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 8]
```diff
@@ -108,4 +130,8 @@ public boolean isSuccess() {
   public String getResponseBody() {
     return responseBody;
   }
+  
+  public Boolean getRetryable() {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: why do we need 2 method - checkRetryable and getRetryable.

Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: why do we need 2 method - checkRetryable and getRetryable.

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 6 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Removed the logger ecfef61ba226b88cb299cac832d75aa7624808ae

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Removed the logger ecfef61ba226b88cb299cac832d75aa7624808ae

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 7 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: Previously this method: `public boolean isSuccess()` was returning true if the HTTP response code was 200 OK.

But there was no logic in place for API calls taking more execution time than the specified limit (60 sec). Here, along with the fetched data, we get an error message saying **Error - com.glide.sys.TransactionCancelledException: Transaction cancelled: maximum execution time exceeded** in the response JSON for all REST API calls that take more than 60 sec and eventually lose that batch. [Data Loss Issue]

To overcome this, we added the retry functionality. Whenever this error is encountered, the connector will retry the API. The `checkRetryable()` method contains the logic as to when the retry functionality should be invoked.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: Previously this method: `public boolean isSuccess()` was returning true if the HTTP response code was 200 OK.

This included the scenario where the ServiceNow Table Rest API returned a 200 OK response for API calls taking more execution time than the specified limit (60 sec). Here, we get an error message saying Error - **com.glide.sys.TransactionCancelledException: Transaction cancelled: maximum execution time exceeded** in the response JSON for all REST API calls that take more than 60 sec and eventually lose that batch. [Data Loss Issue]

To overcome this, we added the retry functionality. Whenever this error is encountered, the connector will retry the API. The `checkRetryable()` method contains the logic as to when the retry functionality should be invoked.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 8 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: The first method : `checkRetryable` is to check whether the retry is required or not. It sets the `isRetryable` variable to true when the error message : _maximum execution time exceeded_ is encountered.
The second method `getRetryable` returns the value of the class level variable: `isRetryable`. It's a getter method.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: The first method : `checkRetryable` is to check whether the retry is required or not. It sets the `isRetryable` variable to true when the error message : _maximum execution time exceeded_ is encountered.
The second method `getRetryable` returns the value of the class level variable: `isRetryable`. It's a getter method.

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 5 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Couldn't find the reference to the library related to retry on the link shared by you.

Can we have a short call for this?

Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Try using https://github.com/rholder/guava-retrying.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 7 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: I see, please add a comment indicating that ServiceNow API is returning OK response with an error msg in response body. Normally we would expect non-200 response code if there's an error

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 9]
```diff
@@ -89,14 +94,20 @@ public String getAccessToken() throws OAuthSystemException, OAuthProblemExceptio
     try {
       String accessToken = getAccessToken();
       requestBuilder.setAuthHeader(accessToken);
+      long start = System.currentTimeMillis();
       apiResponse = executeGet(requestBuilder.build());
+      long end = System.currentTimeMillis();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: start, end are now used. can be removedd

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: `start`, `end` removed. 286b28256112b49eaab4a3a036a86973a3fa3bda

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 7 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java:110]: Added the comment. 76c21ca228efde19affd50f83ad5c0a183bba927

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 5 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Thank you for sharing the library. The dev team is done with the implementation using the guava-retry library and is currently unit testing the code. This will be followed by functional testing by the QA team which will likely take a day or two more to complete. 

As this PR, containing a fix for the data loss issue happens to be a time-sensitive deliverable, part of the ServiceNow Release, can we raise a separate PR for the guava-retry implementation and let this one get through?

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Retry implemented using Guava-retrying library [150737e](https://github.com/data-integrations/servicenow-plugins/pull/16/commits/150737e8afe7681df7a7356fea149043d9c98075)

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 10]
```diff
@@ -216,4 +228,28 @@ private String getErrorMessage(String responseBody) {
       return e.getMessage();
     }
   }
+
+  public List<Map<String, Object>> fetchTableRecordsRetryableMode(String tableName, String startDate, String endDate,
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java:242]: add javadoc for public method

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 11]
```diff
@@ -216,4 +228,28 @@ private String getErrorMessage(String responseBody) {
...
+                                                                  int offset, int limit) {
+    final List<Map<String, Object>> results = new ArrayList<>();
+    Callable<Boolean> fetchRecords = () -> {
+      results.addAll(fetchTableRecords(tableName, startDate, endDate, offset, limit));
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java:246]: fetchTableRecords can now be private

File: pom.xml [hunk 12]
```diff
@@ -215,7 +215,6 @@
       <version>2.24.0</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
```
Comment (rmstar) [inline pom.xml:218]: why was this line removed?

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 13]
```diff
@@ -98,7 +102,27 @@ public int getHttpStatus() {
...
+      isSuccess = false;
+    } else if (successCodes.contains(getHttpStatus())) {
+      isSuccess = true;
+    }
+    return isSuccess;
+  }
+
+  private boolean checkRetryable() {
+    Gson gson = new Gson();
+    JsonObject jo = gson.fromJson(this.getResponseBody(), JsonObject.class);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: getRetryable() will return false incorrectly (for retryble error) if isSuccess() was not called previously on the same RestAPIResponse. We shouldn't have such dependency between getters. You can fix this by moving the checkRetryable logic into the constructor.

File: pom.xml (hunk 12 above)
Comment (sgarg-CS) [reply pom.xml:218]: This one got removed while resolving merge conflicts. 
Added back. https://github.com/data-integrations/servicenow-plugins/pull/16/commits/94cb7501d6219ca8756a6122b6daa510c0b7c1b8

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 10 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java:242]: Added. https://github.com/data-integrations/servicenow-plugins/pull/16/commits/94cb7501d6219ca8756a6122b6daa510c0b7c1b8

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 11 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java:246]: This method is being used by the test classes in a different package.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 13 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Moved the `checkRetryable()` call to the constructor and removed the dependency. https://github.com/data-integrations/servicenow-plugins/pull/16/commits/94cb7501d6219ca8756a6122b6daa510c0b7c1b8

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 14]
```diff
@@ -98,7 +105,27 @@ public int getHttpStatus() {
...
+    // Normally we would expect non-200 response code if there's an error.
+    if (getRetryable()) {
+      throw new RetryableException();
+    } else if (successCodes.contains(getHttpStatus())) {
+      isSuccess = true;
+    }
+    return isSuccess;
+  }
+
+  private boolean checkRetryable() {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Return value is not used, can return void

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 15]
```diff
@@ -98,7 +105,27 @@ public int getHttpStatus() {
   }
 
   public boolean isSuccess() {
-    return successCodes.contains(getHttpStatus());
+    boolean isSuccess = false;
+    // ServiceNow Rest API may return 200 OK response with an error message in the response body.
+    // Normally we would expect non-200 response code if there's an error.
+    if (getRetryable()) {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: no need to call the getter, just access isRetryable directly. Similar comment elsewhere (getResponseBody()).

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Removed the getter : `getRetryable()` 2c4538f24a45de8576bb7a51b2422ad070a696dd

`getResponseBody()` is being used in several places outside the`RestAPIResponse` class

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 14 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Changed the return type to void. 2c4538f24a45de8576bb7a51b2422ad070a696dd

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 16]
```diff
@@ -216,4 +224,39 @@ private String getErrorMessage(String responseBody) {
...
+      return true;
+    };
+
+    Retryer<Boolean> retryer = RetryerBuilder.<Boolean>newBuilder()
+      .retryIfExceptionOfType(RetryableException.class)
+      .withWaitStrategy(WaitStrategies.exponentialWait(120000, TimeUnit.MILLISECONDS))
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Make the wait time and stopAfterAttempt values constants

**Review by rmstar — APPROVED** [review]
lgtm. Please squash commits after addressing minor comment

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 16 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Added constants `WAIT_TIME` and `MAX_NUMBER_OF_RETRY_ATTEMPTS`

**Comment by sgarg-CS** [comment]
Commits have been squashed. Thanks!

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 17]
```diff
@@ -98,7 +104,26 @@ public int getHttpStatus() {
   }
 
   public boolean isSuccess() {
-    return successCodes.contains(getHttpStatus());
+    boolean isSuccess = false;
+    // ServiceNow Rest API may return 200 OK response with an error message in the response body.
+    // Normally we would expect non-200 response code if there's an error.
+    if (this.isRetryable) {
+      throw new RetryableException();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Sorry, I just noticed this. We shouldn't be throwing an exception in this method - just return true/false.
return isRetryable || successCodes.contains(getHttpStatus());

In the caller, if isSuccess() returns false, throw RetryableException if the error is retryable (add a getter for isRetryable)

**Review by rmstar — CHANGES_REQUESTED** [review]
Sorry for the delay on this. You can squash your commits again after addressing the comment.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 17 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Removed the exception from `isSuccess()` method

Added a getter `isRetryable()` in `RestAPIResponse` class. Using this getter to check the value of `isRetryable`  Throwing `RetryableException` if the value is `true`

This is how it looks now:
```
public boolean isSuccess() {
   boolean isSuccess = false;
   // ServiceNow Rest API may return 200 OK response with an error message in the response body.
   // Normally we would expect non-200 response code if there's an error.
   if (this.isRetryable) {
          isSuccess = false;
   } else if (successCodes.contains(getHttpStatus())) {
          isSuccess = true;
   }
   return isSuccess;
}
```

If we use this : `return isRetryable || successCodes.contains(getHttpStatus());`, then when `isRetryable`is `true`, the `isSuccess()` would be true and the below `if` won't get executed and thus retry logic won't be invoked.

```
if (!apiResponse.isSuccess()) {
        if (apiResponse.isRetryable()) {
          throw new RetryableException();
        }
        return Collections.emptyList();
      }
```

**Comment by sgarg-CS** [comment]
Commits squashed. Thanks!

**Review by rmstar — APPROVED** [review]
